### PR TITLE
feat(skills): support metadata-driven runtime config

### LINKED ISSUE #1226: [Feature]: 技能支持metadata 设置appkey能关键信息，openclaw有类似实现
## Summary

[One or two sentences: what do you want and why?]

## Component(s) Affected

- [ ] Core / Backend (app, agents, config, providers, utils, local_models)
- [ ] Console (frontend web UI)
- [ ] Channels (DingTalk, Feishu, QQ, Discord, iMessage, etc.)
- [ ] Skills
- [ ] CLI
- [ ] Documentation (website)
- [ ] Tests
- [ ] CI/CD
- [ ] Scripts / Deploy

## Problem / Motivation

[What problem does this solve? Who benefits?]

## Proposed Solution

[Describe the feature or change you have in mind. Be as specific as possible.]

## Alternatives Considered

[Any other approaches or workarounds you thought about.]

## Additional Context

[Screenshots, examples, links to docs or similar features elsewhere.]

## Willing to Contribute

- [ ] I am willing to open a PR for this feature (after discussion).

### PR DESCRIPTION
**Pull request by fancyboi999**
## Description

This PR turns skill metadata into a usable runtime contract instead of a doc-only field.

It adds structured metadata parsing, skill-level config storage, eligibility checks, and runtime env injection on the backend. It also updates the Console Skills page so users can see metadata, diagnose missing requirements, and edit runtime config without accidentally wiping saved values.

**Related Issue:** Fixes #1226

**Security Considerations:** Skill secrets stay in config instead of `SKILL.md`. Runtime env overrides are applied only for the skill run and restored afterwards.

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation
- [x] Refactoring

## Component(s) Affected

- [x] Core / Backend (app, agents, config, providers, utils, local_models)
- [x] Console (frontend web UI)
- [ ] Channels (DingTalk, Feishu, QQ, Discord, iMessage, etc.)
- [x] Skills
- [ ] CLI
- [ ] Documentation (website)
- [x] Tests
- [ ] CI/CD
- [ ] Scripts / Deploy

## Checklist

- [x] I ran `pre-commit run --all-files` locally and it passes
- [x] If pre-commit auto-fixed files, I committed those changes and reran checks
- [x] I ran tests locally (`pytest` or as relevant) and they pass
- [ ] Documentation updated (if needed)
- [x] Ready for review

## Testing

- Parse `metadata.copaw` / `metadata.openclaw` from `SKILL.md`
- Persist `skills.entries.<skill>.enabled/apiKey/env/config`
- Compute skill eligibility from metadata + config + environment
- Inject `apiKey` and env overrides only for the skill run, then restore them
- Show metadata, eligibility, and runtime config in the Console Skills page
- Load current config into the drawer before editing so saved values are not silently cleared

## Local Verification Evidence

```bash
pre-commit run --all-files
# Passed

PYTHONPATH=src pytest -q tests/test_cli_version.py tests/test_skill_runtime.py
# 3 passed in 9.51s

cd console && npm run build
# Passed
```

## Additional Notes

- Manual browser QA was run against `http://127.0.0.1:8088/skills` to verify card status, drawer rendering, config save, and the preserved-secret flow.
- This is the minimal closed loop for metadata-driven skill config. It does not include installer workflows or broader secret management beyond skill runtime config yet.

### CODE REVIEW

File: console/src/pages/Agent/Skills/components/SkillDrawer.tsx [hunk 1]
```diff
@@ -76,65 +122,138 @@ export function SkillDrawer({
...
-      width={520}
+      width={640}
       placement="right"
       title={editingSkill ? t("skills.viewSkill") : t("skills.createSkill")}
       open={open}
       onClose={onClose}
       destroyOnClose
     >
-      <Form form={form} layout="vertical" onFinish={handleSubmit}>
+      <Form form={form} layout="vertical" onFinish={handleCreateSubmit}>
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply console/src/pages/Agent/Skills/components/SkillDrawer.tsx]: Handled in 34c600b. The edit drawer now disables form submission in edit mode by wiring `onFinish` only for create mode, so pressing Enter can no longer hit the create-skill path.

File: src/copaw/agents/skill_runtime.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+            pending_env = dict(skill_config.env)
+            if metadata and metadata.primary_env and skill_config.api_key:
+                pending_env.setdefault(metadata.primary_env, skill_config.api_key)
+
+            for env_name, env_value in pending_env.items():
+                if not env_name or env_value is None or env_name in os.environ:
+                    continue
+                if env_name not in original_values:
+                    original_values[env_name] = None
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply src/copaw/agents/skill_runtime.py]: Handled in 34c600b. `apply_skill_env_overrides` now records the original value for every injected key, always applies the requested override for the skill run, and restores the previous value afterwards. I also extended the runtime test to cover the override-and-restore path.

File: src/copaw/agents/react_agent.py [hunk 3]
```diff
@@ -534,7 +555,15 @@ async def reply(
...
-        return await super().reply(msg=msg, structured_model=structured_model)
+        eligible_skills = self._sync_registered_skills(self.toolkit)
+        with apply_skill_env_overrides(
+            skills=eligible_skills,
+            config=load_config(),
+        ):
+            return await super().reply(
+                msg=msg,
+                structured_model=structured_model,
+            )
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply src/copaw/agents/react_agent.py:561]: Handled in 34c600b. I added a module-level async lock around the env override window in `ReactAgent.reply`, and only enter that locked path when the current skill set actually needs env/API-key injection. That keeps the current design safe against concurrent request leakage while preserving the existing skill execution flow.

File: console/src/pages/Agent/Skills/components/SkillDrawer.tsx [hunk 4]
```diff
@@ -170,11 +282,75 @@ export function SkillDrawer({
...
+                </div>
+              </div>
+              <div
+                className={`${styles.eligibilityBadge} ${
+                  editingSkill.eligibility?.eligible
+                    ? styles.eligible
+                    : styles.ineligible
+                }`}
+              >
+                {editingSkill.eligibility?.eligible
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply console/src/pages/Agent/Skills/components/SkillDrawer.tsx]: Handled in 34c600b. The drawer badge now defaults to eligible when `eligibility` is absent, so the detail view stays consistent with the card and transitional states no longer show a false negative.

File: console/src/pages/Agent/Skills/components/SkillCard.tsx [hunk 5]
```diff
@@ -138,9 +143,16 @@ export function SkillCard({
...
+          <code className={styles.infoCode}>
+            {isEligible ? t("skills.eligible") : t("skills.ineligible")}
+          </code>
+        </div>
+
+        <div className={styles.infoSection}>
+          <div className={styles.infoLabel}>{t("skills.skillKey")}</div>
           <code className={`${styles.infoCode} ${styles.path}`}>
-            {skill.path}
+            {skill.resolved_skill_key || skill.path}
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply console/src/pages/Agent/Skills/components/SkillCard.tsx]: Handled in 34c600b. The Skill Key field now falls back to `skill.name` instead of `skill.path`, which keeps the label semantically correct and avoids leaking filesystem paths in the UI.

File: src/copaw/agents/skill_runtime.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+                os.environ[env_name] = env_value
+
+        yield
+    finally:
+        for env_name in reversed(list(original_values.keys())):
+            original = original_values[env_name]
+            if original is None:
+                os.environ.pop(env_name, None)
+            else:
+                os.environ[env_name] = original
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply src/copaw/agents/skill_runtime.py]: Handled in 34c600b. I kept the env-based runtime design for compatibility, but the mutation window is now guarded by a module-level async lock so concurrent requests cannot observe each other's temporary env values. The overrides are still restored in `finally`, and the runtime test now covers restore semantics.

File: console/src/pages/Agent/Skills/components/SkillDrawer.tsx [hunk 7]
```diff
@@ -76,65 +122,138 @@ export function SkillDrawer({
...
+    loadConfig().catch(() => {
+      if (!cancelled) {
+        setLoadingConfig(false);
+      }
+    });
+
+    return () => {
+      cancelled = true;
+    };
+  }, [editingSkill, form, onLoadConfig]);
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply console/src/pages/Agent/Skills/components/SkillDrawer.tsx:199]: Handled in 34c600b. The drawer now tracks config-load failure explicitly, resets that state on reopen, and disables Save when the current config could not be loaded. That blocks the empty-overwrite path you called out.

File: src/copaw/app/routers/skills.py [hunk 8]
```diff
@@ -189,6 +217,67 @@ async def create_skill(request: CreateSkillRequest):
...
+    else:
+        config.skills.entries[skill_key] = existing
+    save_config(config)
+
+    refreshed = SkillService.get_skill(skill_name)
+    refreshed_key = refreshed.resolved_skill_key if refreshed else skill_key
+    return _build_skill_config_view(
+        refreshed_key or skill_key,
+        config.skills.entries.get(refreshed_key or skill_key),
+    )
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply src/copaw/app/routers/skills.py:329]: Handled in 34c600b. `PUT /skills/{skill_name}/config` now validates `env` keys against the skill's declared metadata contract (`primaryEnv` plus `requires.env`) before persisting anything. This closes the arbitrary-env injection path and keeps skill runtime config scoped to metadata-declared inputs.

File: src/copaw/agents/skill_runtime.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+                    metadata.primary_env,
+                    skill_config.api_key,
+                )
+
+            for env_name, env_value in pending_env.items():
+                if not env_name or env_value is None:
+                    continue
+                if env_name not in original_values:
+                    original_values[env_name] = os.environ.get(env_name)
+                os.environ[env_name] = env_value
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply src/copaw/agents/skill_runtime.py:207]: Handled in 402e45d. `apply_skill_env_overrides` now detects conflicting values for the same env key across active skills and raises a clear error instead of letting iteration order silently decide which secret wins.

[the remark this answers is not in the document]
Comment (fancyboi999) [reply src/copaw/agents/skill_runtime.py]: Handled in 402e45d. Runtime injection now reuses the metadata-derived allowlist and filters `SkillEntryConfig.env` before building the pending env map, so manual or legacy config edits cannot bypass the declared skill contract.

File: console/src/pages/Agent/Skills/components/SkillDrawer.tsx [hunk 10]
```diff
@@ -76,65 +123,152 @@ export function SkillDrawer({
...
+      const env = parseJsonRecord(values.envJson || "", "env");
+      const config = parseJsonRecord(values.configJson || "", "config");
+      const payload: SkillConfigUpdatePayload = {
+        enabled: values.skillEnabled ?? null,
+        clearApiKey: values.clearApiKey || false,
+        env: Object.fromEntries(
+          Object.entries(env).map(([key, value]) => [key, String(value ?? "")]),
+        ),
+        config,
+      };
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply console/src/pages/Agent/Skills/components/SkillDrawer.tsx]: Handled in 402e45d. The drawer now makes the enabled override explicit with a separate `Override runtime enabled state` switch. When that override is off, the payload sends `enabled: null`, so users can return to inherit mode instead of being forced into a persistent boolean override.

File: src/copaw/app/routers/skills.py [hunk 11]
```diff
@@ -59,44 +80,73 @@ class HubInstallRequest(BaseModel):
...
+def _build_skill_config_view(
+    skill_key: str,
+    entry: SkillEntryConfig | None,
+) -> SkillConfigView:
+    entry = entry or SkillEntryConfig()
+    return SkillConfigView(
+        key=skill_key,
+        enabled=entry.enabled,
+        has_api_key=bool(entry.api_key),
+        env=dict(entry.env or {}),
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply src/copaw/app/routers/skills.py]: Handled in 4c2f330. `SkillConfigView` now masks env values before returning them to the client, so the config API no longer round-trips raw env secrets. The update path also preserves existing masked values when the user leaves them unchanged, which keeps the UI editable without leaking the original secret values back over the API.

File: src/copaw/agents/skill_runtime.py [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+
+def _resolve_config_path(config: Config | None, path: str) -> Any:
+    current: Any = (
+        config.model_dump(mode="json", by_alias=True) if config else {}
+    )
+    for part in [item for item in path.split(".") if item]:
+        if not isinstance(current, dict):
+            return None
+        current = current.get(part)
+    return current
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply src/copaw/agents/skill_runtime.py]: Handled in 4c2f330. `compute_skill_eligibility` now serializes the config once per call and reuses that dumped structure when resolving multiple `requirements.config` paths, so we no longer pay for repeated full-config `model_dump()` calls inside `_resolve_config_path`.

**Comment by Leirunlin** [comment]
Hi! @fancyboi999, could you please help resolve the conflicts with the latest copaw?

File: src/copaw/config/config.py [hunk 13]
```diff
@@ -442,6 +442,39 @@ class ToolsConfig(BaseModel):
     )
 
 
+class SkillEntryConfig(BaseModel):
+    """Configuration for a single skill."""
+
+    model_config = ConfigDict(populate_by_name=True)
+
+    enabled: Optional[bool] = Field(
+        default=None,
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply src/copaw/config/config.py:457]: Handled in d3e4a4c. `SkillEntryConfig` now normalizes `apiKey: null` to an empty string before validation, so legacy or hand-edited config entries no longer break `load_config()`. I also added a regression test covering the null-input path.

File: src/copaw/agents/skills_manager.py [hunk 14]
```diff
@@ -531,7 +574,30 @@ def list_available_skills() -> list[SkillInfo]:
         Returns:
             List of SkillInfo with name, content, source, and path.
         """
-        return _read_skills_from_dir(get_active_skills_dir(), "active")
+        return _read_skills_from_dir(
+            get_active_skills_dir(),
+            "active",
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply src/copaw/agents/skills_manager.py:590]: Handled in d3e4a4c. `SkillService.get_skill()` no longer scans every skill directory via `_read_skills_from_dir(...)`; it now resolves `{directory}/{name}` directly and parses only that skill. I also added a regression test that fails if `get_skill()` falls back to the old full-directory scan.

File: src/copaw/app/routers/skills.py [hunk 15]
```diff
@@ -59,34 +81,90 @@ class HubInstallRequest(BaseModel):
...
+    existing_entry: SkillEntryConfig,
+    env_payload: dict[str, str] | None,
+) -> dict[str, str]:
+    env_payload = env_payload or {}
+    merged_env: dict[str, str] = {}
+    for key, value in env_payload.items():
+        if value == MASKED_ENV_VALUE and key in existing_entry.env:
+            merged_env[key] = existing_entry.env[key]
+        else:
+            merged_env[key] = value
```
[the remark this answers is not in the document]
Comment (fancyboi999) [reply src/copaw/app/routers/skills.py:159]: Handled in d3e4a4c. `_merge_skill_env_payload` now rejects the masked sentinel for newly introduced env keys and returns a 400 unless that key already has a stored value to preserve. That prevents persisting `*****` as if it were a real secret, and the router test now covers this case.

**Comment by fancyboi999** [comment]
I also ran a local post-merge smoke check against the latest branch state.

Validation summary:
- Resolved the latest `origin/main` merge conflicts and re-ran the targeted checks.
- Ran `PYTHONPATH=src pytest -q tests/test_skill_runtime.py` and `pre-commit` on the touched backend files.
- Built the console and opened `http://127.0.0.1:8088/skills` locally in a browser.
- Verified the Skills page renders correctly after the merge, opened the `browser_visible` drawer, and confirmed metadata / requirements / runtime config sections render as expected.
- Performed a real UI save through the drawer (`Extra Config JSON`), observed the success toast, verified the persisted value through `GET /api/skills/browser_visible/config`, then cleaned the test value back out.
